### PR TITLE
[CN-445] Add support for custom class from configmaps

### PR DESCRIPTION
**Pull request by mtyazici**
Changes:
- Added ConfigMap support to Custom Class feature.
- Refactored StatefulSet reconciliation function to centralize creation of `initContainer`, `volumes` and `volumeMounts` fields. This allows us to directly use these functions in the `createOrUpdate` function without any if-else check in the `createOrUpdate` function.
- Reordered functions

### CODE REVIEW

File: controllers/hazelcast/hazelcast_resources.go [hunk 1]
```diff
@@ -783,6 +783,17 @@ func backupAgentContainer(h *hazelcastv1alpha1.Hazelcast) v1.Container {
 	}
 }
 
+func initContainers(h *hazelcastv1alpha1.Hazelcast) []corev1.Container {
+	var ctnrs []corev1.Container
```
Comment (dzeromski-hazelcast) [inline controllers/hazelcast/hazelcast_resources.go:787]: Why `ctnrs` and not just  `containers`? :wink:

File: controllers/hazelcast/hazelcast_resources.go [hunk 2]
```diff
@@ -1026,6 +1061,19 @@ func env(h *hazelcastv1alpha1.Hazelcast) []v1.EnvVar {
 	return envs
 }
 
+func javaClassPath(h *hazelcastv1alpha1.Hazelcast) string {
+	b := []string{n.CustomClassBucketPath + "/*"}
+	if !h.Spec.CustomClass.IsConfigMapEnabled() {
```
Comment (dzeromski-hazelcast) [inline controllers/hazelcast/hazelcast_resources.go:1066]: I think this is not needed: `strings.Join([]string{"a"}, ":")` returns `a` (https://go.dev/play/p/YWtbx-aUu4Y).

**Comment by dzeromski-hazelcast** [comment]
Great work with refactoring!

File: api/v1alpha1/hazelcast_types.go [hunk 3]
```diff
@@ -280,8 +287,13 @@ func (c *ExposeExternallyConfiguration) IsEnabled() bool {
 }
 
 // Returns true if customClass configuration is specified.
-func (c *CustomClassConfiguration) IsEnabled() bool {
-	return c != nil && !(*c == (CustomClassConfiguration{}))
+func (c *CustomClassConfiguration) IsBucketEnabled() bool {
+	return c != nil && c.BucketConfiguration != nil
+}
+
+// Returns true if customClass configuration is specified.
```
Comment (cagric0) [inline api/v1alpha1/hazelcast_types.go:294]: A clearer comment would be better.

**Review by cagric0 — COMMENTED** [review]
LGTM except one issue. What if BucketConfiguration and Configmaps are given under CustomClassConfiguration at the same time? I think we are not checking it and It can cause some problems.

File: controllers/hazelcast/hazelcast_resources.go (hunk 2 above)
Comment (mtyazici) [reply controllers/hazelcast/hazelcast_resources.go:1066]: I am not sure what is not needed, after the if check we are adding into the slice for configmaps. Can you share a snippet of how the function should change?

**Comment by mtyazici** [comment]
@cagric0, users can give both configmaps and bucket configurations  at the same time, they are not mutually exclusive. So we do not need to check for it.

File: api/v1alpha1/hazelcast_types.go (hunk 3 above)
Comment (mtyazici) [reply api/v1alpha1/hazelcast_types.go:294]: Sure, I will update it.

**Review by cagric0 — APPROVED** [review]

File: controllers/hazelcast/hazelcast_resources.go (hunk 2 above)
Comment (dzeromski-hazelcast) [reply controllers/hazelcast/hazelcast_resources.go:1066]: Ah now I see you negate first condition. Well then maybe just remove: `if !h.Spec.CustomClass.IsConfigMapEnabled() {` condition?

**Review by dzeromski-hazelcast — APPROVED** [review]

**Comment by dzeromski-hazelcast** [comment]
Just to clarify do we need to worry about CustomClass.TriggerSequence in case of configmaps?

**Comment by mtyazici** [comment]
Files are updated by kubelet but Hazelcast won't pick up the changes because the classes are loaded at the startup AFAIK.

File: controllers/hazelcast/hazelcast_resources.go (hunk 2 above)
Comment (mtyazici) [reply controllers/hazelcast/hazelcast_resources.go:1065]: When we remove the condition it results in nil pointer dereference

**Comment by mtyazici** [comment]
Flaky test not related to the PR failed, I will merge the PR.

**Pull request merged**